Support Apple m1

changvvb opened the pull request:
Support Apple m1

https://github.com/ClickHouse/ClickHouse/issues/20245

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
 - Support Apple m1
 
Known issues:
 - Rocksdb engine does not work because of the wild pointer problem (**DISABLED**).
 - Debug mode binary run failed because two asserts failed in submodule `jemalloc` (**DISABLED**).

      contrib/jemalloc/include/jemalloc/internal/tsd.h:330
      ```cpp
           assert(!malloc_slow && tsd_tcache_enabled_get(tsd) &&
               tsd_reentrancy_level_get(tsd) == 0);
      ```
      
      contrib/jemalloc/include/jemalloc/internal/tsd_generic.h:145
      ```cpp
           	assert(tsd_booted);
      ```

Commit:
Support apple m1

Commit:
Update

Commit:
Update

Commit:
revert submodule commit

Commit:
Update submodule url

Commit:
try to fix rocksdb but not success

Commit:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into support-apple-m1

Commit:
Add  cmake/darwin/toolchain-aarch64.cmake

Commit:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into support-apple-m1

Commit:
Update

Commit:
Update

Commit:
Update

Commit:
Update

Commit:
Update .gitmodules

Commit:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into support-apple-m1

yanhuqing666 commented:
Hi,  I found a mistake in the Known issues:

```
contrib/jemalloc/include/jemalloc/internal/tsd_generic.h:145

     assert(!malloc_slow && tsd_tcache_enabled_get(tsd) &&
         tsd_reentrancy_level_get(tsd) == 0);
```

It should be  
```
contrib/jemalloc/include/jemalloc/internal/tsd_generic.h:145

     assert(tsd_booted);
```

And when I add comments on these lines,  Debug mode binary can only run for the first time.  
Do you have any idea about debugging on Apple M1?

changvvb commented:
> Hi, I found a mistake in the Known issues:
> 
> ```
> contrib/jemalloc/include/jemalloc/internal/tsd_generic.h:145
> 
>      assert(!malloc_slow && tsd_tcache_enabled_get(tsd) &&
>          tsd_reentrancy_level_get(tsd) == 0);
> ```
> 
> It should be
> 
> ```
> contrib/jemalloc/include/jemalloc/internal/tsd_generic.h:145
> 
>      assert(tsd_booted);
> ```
> 
> And when I add comments on these lines, Debug mode binary can only run for the first time.
> Do you have any idea about debugging on Apple M1?

Thanks for your report. I have updated the description.
After commenting the lines, debug mode binary can run on my machine without other issues. What problem did you face?
On the other hand, release mode binary can also be debugged by llvm, but maybe lose some information.

yanhuqing666 commented:
> After commenting the lines, debug mode binary can run on my machine without other issues. What problem did you face?
> On the other hand, release mode binary can also be debugged by llvm, but maybe lose some information.

When system tables merge, it will occur Invalid address alignment:


> 2021.03.26 14:15:31.909083 [ 30240 ] {} <Debug> system.metric_log (d4724d59-77b7-4902-a695-65160a87e193) (MergerMutator): Merging 8 parts: from 202103_1_1_0 to 202103_8_8_0 into Compact
> 2021.03.26 14:15:31.911141 [ 30240 ] {} <Debug> system.metric_log (d4724d59-77b7-4902-a695-65160a87e193) (MergerMutator): Selected MergeAlgorithm: Horizontal
> 2021.03.26 14:15:31.922158 [ 30240 ] {} <Trace> MergeTreeSequentialSource: Reading 2 marks from part 202103_1_1_0, total 8 rows starting from the beginning of the part
> 
> 2021.03.26 14:15:31.942960 [ 30240 ] {} <Trace> MergeTreeSequentialSource: Reading 2 marks from part 202103_2_2_0, total 8 rows starting from the beginning of the part
> 
> 2021.03.26 14:15:31.949565 [ 30217 ] {} <Trace> BaseDaemon: Received signal 10
> 2021.03.26 14:15:31.949859 [ 30249 ] {} <Fatal> BaseDaemon: ########################################
> 2021.03.26 14:15:31.949943 [ 30249 ] {} <Fatal> BaseDaemon: (version 21.3.1.1, no build id) (from thread 30240) (no query) Received signal Bus error: 10 (10)
> 2021.03.26 14:15:31.949995 [ 30249 ] {} <Fatal> BaseDaemon: Invalid address alignment.
> 2021.03.26 14:15:31.950024 [ 30249 ] {} <Fatal> BaseDaemon: Stack trace: 0x1980596cc
> 2021.03.26 14:15:31.950261 [ 30249 ] {} <Fatal> BaseDaemon: 0. 0x1980596cc

changvvb commented:
> When system tables merge, it will occur Invalid address alignment:
> 
> >
@yanhuqing666  Would you mind providing the SQL in your example?

yanhuqing666 commented:
> > When system tables merge, it will occur Invalid address alignment:
> > >
> 
> @yanhuqing666 Would you mind providing the SQL in your example?

No client connect to the server, just start for a while and flushing system log

/mypath/workspace/ClickHouse/cmake-build-debug/programs/clickhouse-server --config-file=/mypath/workspace/ClickHouse/programs/server/config.xml

I delete these file
 ../programs/server/config.d/test_keeper_port.xml
 ../programs/server/config.d/zookeeper.xml

changvvb commented:
> 
> When system tables merge, it will occur Invalid address alignment:
> 

@yanhuqing666 I tested it on my machine and merged successfully. Maybe we need more information to reproduce.

> 2021.03.26 16:27:53.736202 [ 6674041 ] {} <Debug> default.test_merge_tree (ce9b36a7-708d-44c1-8a69-bcf4f4d97311) (MergerMutator): Selected MergeAlgorithm: Horizontal
2021.03.26 16:27:53.736567 [ 6674041 ] {} <Trace> MergeTreeSequentialSource: Reading 2 marks from part all_1_85_17, total 85 rows starting from the beginning of the part
2021.03.26 16:27:53.737508 [ 6674041 ] {} <Trace> MergeTreeSequentialSource: Reading 2 marks from part all_86_86_0, total 1 rows starting from the beginning of the part
2021.03.26 16:27:53.738355 [ 6674041 ] {} <Trace> MergeTreeSequentialSource: Reading 2 marks from part all_87_87_0, total 1 rows starting from the beginning of the part
2021.03.26 16:27:53.739250 [ 6674041 ] {} <Trace> MergeTreeSequentialSource: Reading 2 marks from part all_88_88_0, total 1 rows starting from the beginning of the part
2021.03.26 16:27:53.740115 [ 6674041 ] {} <Trace> MergeTreeSequentialSource: Reading 2 marks from part all_89_89_0, total 1 rows starting from the beginning of the part
2021.03.26 16:27:53.740979 [ 6674041 ] {} <Trace> MergeTreeSequentialSource: Reading 2 marks from part all_90_90_0, total 1 rows starting from the beginning of the part
2021.03.26 16:27:53.751529 [ 6674041 ] {} <Debug> default.test_merge_tree (ce9b36a7-708d-44c1-8a69-bcf4f4d97311) (MergerMutator): Merge sorted 90 rows, containing 2 columns (2 merged, 0 gathered) in 0.015588 sec., 5773.6720554272515 rows/sec., 118.41 KiB/sec.
2021.03.26 16:27:53.753502 [ 6674041 ] {} <Trace> default.test_merge_tree (ce9b36a7-708d-44c1-8a69-bcf4f4d97311): Renaming temporary part tmp_merge_all_1_90_18 to all_1_90_18.
2021.03.26 16:27:53.754210 [ 6674041 ] {} <Trace> default.test_merge_tree (ce9b36a7-708d-44c1-8a69-bcf4f4d97311) (MergerMutator): Merged 6 parts: from all_1_85_17 to all_90_90_0
>

Commit:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into support-apple-m1

yanhuqing666 commented:
I delete these file 

> ../programs/server/config.d/metric_log.xml
> ../programs/server/config.d/text_log.xml

and I add comments on these lines in config.xml:

>     <metric_log>
>         <database>system</database>
>         <table>metric_log</table>
>         <flush_interval_milliseconds>7500</flush_interval_milliseconds>
>         <collect_interval_milliseconds>1000</collect_interval_milliseconds>
>     </metric_log>

Now ch works!
But I don't konw why...

Commit:
Update jemalloc_internal_defs & fix unit test

traceon commented:
Great job! Maybe it is worth just disabling jemalloc and rocksdb support in Apple silicon target, and make this PR ready? Those then could be addressed in further iterations, I guess.

Commit:
Fix unit test

Commit:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into support-apple-m1

Commit:
Disabled rocksdb jemalloc

changvvb commented on src/IO/tests/gtest_DateTimeToString.cpp:
```diff
@@ -161,7 +161,7 @@ INSTANTIATE_TEST_SUITE_P(DateTimeToString, DateTimeToStringParamTestDayNum,
             "Negative DayNum value wraps as if it was UInt16 due to LUT limitations and to maintain compatibility with existing code.",
             DayNum(-10 * 365),
-            "2106-02-07"
+            "2139-06-10"
```
I don't know why the expected result was `"2106-02-07"`
```cpp
#include <iostream>
#include <string>

int main()
{
  unsigned short t = -10 * 365;
  
  std::cout << t << std::endl;                  // 61886
  std::cout << t / 365 << std::endl;            // 169
  std::cout << t / 365 + 1970 << std::endl;     // 2139
}
```

Commit:
GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST(DateLUTWithTimeZoneAndTimeRange);

Commit:
revert rocksdb submodule

changvvb commented:
I think it's ready to review this PR.

mishfish commented:
hi @changvvb 
how do you build ClickHouse?
I try with
`cmake -DCMAKE_C_COMPILER=$(brew --prefix gcc)/bin/gcc-10 -DCMAKE_CXX_COMPILER=$(brew --prefix gcc)/bin/g++-10 -DCMAKE_BUILD_TYPE=RelWithDebInfo -DENABLE_JEMALLOC=OFF ..`
but it doesn't build

Have I use some toolchain for build?

changvvb commented:
> how do you build ClickHouse?
>
Hi @michon470, you should use `clang` to build the project.
Try this
```sh
cmake .. -DCMAKE_C_COMPILER=`brew --prefix llvm`/bin/clang -DCMAKE_CXX_COMPILER=`brew --prefix llvm`/bin/clang++ -DCMAKE_PREFIX_PATH=`brew --prefix llvm` -DENABLE_JEMALLOC=OFF
```

mishfish commented:
@changvvb I've been trying, but got only

`CMake Error at cmake/arch.cmake:14 (message):
  32bit platforms are not supported
Call Stack (most recent call first):
  CMakeLists.txt:44 (include)`

changvvb commented:
> @changvvb I've been trying, but got only
> 
> `CMake Error at cmake/arch.cmake:14 (message): 32bit platforms are not supported Call Stack (most recent call first): CMakeLists.txt:44 (include)`

According to  https://github.com/ClickHouse/ClickHouse/pull/21639/files#diff-d5d8fc115656969af2961dbe1b6bf7f8a722df4998be94e0936d8be0dee4f733 , I think you should check `CMAKE_SYSTEM_PROCESSOR`, `ARCH_AARCH64`, `ARCH_ARM`.

buyology commented:
Really nice work 👍 

> ```shell
> cmake .. -DCMAKE_C_COMPILER=`brew --prefix llvm`/bin/clang -DCMAKE_CXX_COMPILER=`brew --prefix llvm`/bin/clang++ -DCMAKE_PREFIX_PATH=`brew --prefix llvm` -DENABLE_JEMALLOC=OFF
> ```

Adding a data point: ✅ Tested compiling with the above, and can confirm that I got it to work.

mishfish commented:
Unfortunately still cannot build it:)

This time I get

`[2476/5628] Building ASM object contrib/boost-cmake/CMakeFiles/_boost_context.dir/__/boost/libs/context/src/asm/jump_arm64_aapcs_elf_gas.S.o
FAILED: contrib/boost-cmake/CMakeFiles/_boost_context.dir/__/boost/libs/context/src/asm/jump_arm64_aapcs_elf_gas.S.o
/opt/homebrew/opt/llvm/bin/clang  -I../contrib/boost -isystem ../contrib/libcxx/include -isystem ../contrib/libcxxabi/include -O2 -g -DNDEBUG -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk -mmacosx-version-min=10.15   -D OS_DARWIN -MD -MT contrib/boost-cmake/CMakeFiles/_boost_context.dir/__/boost/libs/context/src/asm/jump_arm64_aapcs_elf_gas.S.o -MF contrib/boost-cmake/CMakeFiles/_boost_context.dir/__/boost/libs/context/src/asm/jump_arm64_aapcs_elf_gas.S.o.d -o contrib/boost-cmake/CMakeFiles/_boost_context.dir/__/boost/libs/context/src/asm/jump_arm64_aapcs_elf_gas.S.o -c ../contrib/boost/libs/context/src/asm/jump_arm64_aapcs_elf_gas.S
../contrib/boost/libs/context/src/asm/jump_arm64_aapcs_elf_gas.S:58:1: error: unknown directive
.type jump_fcontext, %function
^
../contrib/boost/libs/context/src/asm/jump_arm64_aapcs_elf_gas.S:112:1: error: unknown directive
.size jump_fcontext,.-jump_fcontext
^
../contrib/boost/libs/context/src/asm/jump_arm64_aapcs_elf_gas.S:114:19: error: unexpected token in '.section' directive
.section .note.GNU-stack,"",%progbits
                  ^
[2477/5628] Building ASM object contrib/boost-cmake/CMakeFiles/_boost_context.dir/__/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S.o
FAILED: contrib/boost-cmake/CMakeFiles/_boost_context.dir/__/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S.o
/opt/homebrew/opt/llvm/bin/clang  -I../contrib/boost -isystem ../contrib/libcxx/include -isystem ../contrib/libcxxabi/include -O2 -g -DNDEBUG -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk -mmacosx-version-min=10.15   -D OS_DARWIN -MD -MT contrib/boost-cmake/CMakeFiles/_boost_context.dir/__/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S.o -MF contrib/boost-cmake/CMakeFiles/_boost_context.dir/__/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S.o.d -o contrib/boost-cmake/CMakeFiles/_boost_context.dir/__/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S.o -c ../contrib/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S
../contrib/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S:58:1: error: unknown directive
.type make_fcontext, %function
^
../contrib/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S:83:1: error: unknown directive
.size make_fcontext,.-make_fcontext
^
../contrib/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S:85:19: error: unexpected token in '.section' directive
.section .note.GNU-stack,"",%progbits
                  ^
[2485/5628] Building CXX object contrib/antlr4-runtime-cmake/CMakeFiles/antlr4-runtime.dir/__/antlr4-runtime/UnbufferedCharStream.cpp.o
ninja: build stopped: subcommand failed.`

Try to investigate this issue

buyology commented:
> Unfortunately still cannot build it:)
> 
> This time I get
> 
> `
> […] ../contrib/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S:58:1: error: unknown directive .type make_fcontext, %function ^ ../contrib/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S:83:1: error: unknown directive .size make_fcontext,.-make_fcontext ^ ../contrib/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S:85:19: error: unexpected token in '.section' directive .section .note.GNU-stack,"",%progbits ^ [2485/5628] Building CXX object contrib/antlr4-runtime-cmake/CMakeFiles/antlr4-runtime.dir/__/antlr4-runtime/UnbufferedCharStream.cpp.o ninja: build stopped: subcommand failed.
> `

This sounds precisely like what the PR at hand is patching<sup>[1]</sup>. I would recommend a fresh clone from the <kbd>changvvb:support-apple-m1</kbd>-branch just to double check that you are building the right things with the right configuration.

<sup>[1]</sup> https://github.com/ClickHouse-Extras/boost/blob/a8d43d3142cc6b26fc55bec33f7f6edb1156ab7a/libs/context/src/asm/make_arm64_aapcs_elf_gas.S#L85-L89

mishfish commented:
> > Unfortunately still cannot build it:)
> > This time I get
> > `[…] ../contrib/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S:58:1: error: unknown directive .type make_fcontext, %function ^ ../contrib/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S:83:1: error: unknown directive .size make_fcontext,.-make_fcontext ^ ../contrib/boost/libs/context/src/asm/make_arm64_aapcs_elf_gas.S:85:19: error: unexpected token in '.section' directive .section .note.GNU-stack,"",%progbits ^ [2485/5628] Building CXX object contrib/antlr4-runtime-cmake/CMakeFiles/antlr4-runtime.dir/__/antlr4-runtime/UnbufferedCharStream.cpp.o ninja: build stopped: subcommand failed.`
> 
> This sounds precisely like what the PR at hand is patching[1]. I would recommend a fresh clone from the changvvb:support-apple-m1-branch just to double check that you are building the right things with the right configuration.
> 
> [1] https://github.com/ClickHouse-Extras/boost/blob/a8d43d3142cc6b26fc55bec33f7f6edb1156ab7a/libs/context/src/asm/make_arm64_aapcs_elf_gas.S#L85-L89

Thanks a lot, you are right

mishfish commented:
@buyology @changvvb 
Thank you a lot, I finally compile it:)
Sorry for stupid questions:)

Commit:
Update openldap header files

traceon commented:
@changvvb This commit doesn't look good: https://github.com/ClickHouse/ClickHouse/pull/21639/commits/21a2eb32f26496692cef118c8a04a0e20fbc1194
It doesn't follow the same feature check pattern as for other platforms.

traceon committed:
Disable jemalloc for ALL macOS non-RelWithDebInfo builds
Revisit default ENABLE_* flag setting

traceon committed:
Remove -DENABLE_JEMALLOC=OFF everywhere as the default value should work
Remove extra '$' in the bach command list to ease copy-pasteing
Adjust comments for alternatives in commands
Add note about ARM (Apple Silicon) support

traceon committed:
Syntax fix

traceon commented:
@changvvb could you please sync with the master?

Commit:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into support-apple-m1

changvvb commented:
@traceon I have merge master to this PR.  As for ldap, maybe I need some hours to hack it.

traceon commented:
@changvvb do you experience any specific problem with it? I believe, the same set of config header files should work for both x86_64 and arm64.

changvvb commented:
> do you experience any specific problem with it?

No. I just execute `./configure` at `contrib/openldap` and copy the generated files to `contrib/openldap-cmake`.
Maybe I lose some parameters of `./configure`, I guess.

traceon commented:
Ok, so then let's use the copy-pasted one (I reverted to it in https://github.com/changvvb/ClickHouse/pull/1 ) - it works for me, everything compiles.

I'll try to separately look into GCC build (Clang and AppleClang even with Xcode IDE after some zlib-ng patch that hopefully will be merged soon) work with that small changes from that PR.

traceon committed:
ARM to arm64

traceon committed:
Bump zlib-ng with a fix for arm64 AppleClang linker

changvvb commented:
> Ok, so then let's use the copy-pasted one (I reverted to it in [changvvb#1](https://github.com/changvvb/ClickHouse/pull/1) ) - it works for me, everything compiles.

OK, should I also revert that commit in this PR?

Commit:
Revert "Update openldap header files"

This reverts commit 21a2eb32f26496692cef118c8a04a0e20fbc1194.

traceon commented:
No, I am trying to commit those changes in your branch actually, so they will be reverted there. If you could merge my changes, that would be great.

traceon commented:
Let me prepare a new PR to your branch.

traceon committed:
Merge pull request #2844 from changvvb/support-apple-m1

wip

changvvb commented:
Yes, I will merge it after you complete the job.

traceon commented:
@changvvb could you please merge this: https://github.com/changvvb/ClickHouse/pull/2

changvvb committed:
Merge pull request #2 from traceon/support-apple-m1-changvvb-2

Support apple m1: proposed adjustments to the original M1 support PR branch

traceon commented:
Great, so now AppleClang with Xcode or Ninja should work too, just as described in the docs.

Commit:
rebuild

Commit:
Merge branch 'support-apple-m1' of github.com:changvvb/ClickHouse into support-apple-m1

Commit:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into support-apple-m1

traceon commented on contrib/jemalloc-cmake/CMakeLists.txt:
```diff
@@ -1,10 +1,13 @@
-if (SANITIZE OR NOT (ARCH_AMD64 OR ARCH_ARM OR ARCH_PPC64LE) OR NOT (OS_LINUX OR OS_FREEBSD OR OS_DARWIN))
+if (SANITIZE OR NOT (
+    ((OS_LINUX OR OS_FREEBSD) AND (ARCH_AMD64 OR ARCH_ARM OR ARCH_PPC64LE)) OR
+    (OS_DARWIN AND ARCH_AMD64 AND CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")
```
If jemalloc is expected to work with `RelWithDebInfo` arm64 macOS builds too, then it makes sense to change to this:
```suggestion
    (OS_DARWIN AND CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo")
```

traceon commented on contrib/jemalloc-cmake/CMakeLists.txt:
```diff
@@ -1,10 +1,13 @@
         message (${RECONFIGURE_MESSAGE_LEVEL}
-                 "jemalloc is disabled implicitly: it doesn't work with sanitizers and can only be used with x86_64, aarch64 or ppc64le on linux or freebsd.")
-    endif()
+                 "jemalloc is disabled implicitly: it doesn't work with sanitizers and can only be used with x86_64, aarch64, or ppc64le Linux or FreeBSD builds and x86_64 RelWithDebInfo macOS builds.")
```
...and adjust this:
```suggestion
                 "jemalloc is disabled implicitly: it doesn't work with sanitizers and can only be used with x86_64, aarch64, or ppc64le Linux or FreeBSD builds and RelWithDebInfo macOS builds.")
```

changvvb committed:
Apply suggestions from code review

Co-authored-by: Denis Glazachev <traceon@users.noreply.github.com>

traceon commented:
@changvvb could you please sync with `master` one more time?

Commit:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into support-apple-m1

changvvb commented:
> @changvvb could you please sync with `master` one more time?

Done

traceon committed:
Merge pull request #2852 from changvvb/support-apple-m1

Sync

traceon committed:
Remove unneeded changes
Fix GCC build

traceon commented:
Thanks, @changvvb . Could you please also merge this: https://github.com/changvvb/ClickHouse/pull/3 . This is my final change, hopefully.

changvvb committed:
Merge pull request #3 from traceon/support-apple-m1-changvvb-3

Support apple m1: proposed adjustments to the original M1 support PR branch

traceon commented:
Performance numbers are published here: https://clickhouse.tech/benchmark/hardware/ (search for "MacBook Air M1".)

changvvb commented:
The performance on MacBook Air M1 seems even better than MacBook Pro 2020.

kitaisreal committed:
Updated zlib version

Commit:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into support-apple-m1

Commit:
Merge branch 'support-apple-m1' of github.com:changvvb/ClickHouse into support-apple-m1

kitaisreal committed:
Updated zlib version

kitaisreal committed:
Merge branch 'master' into support-apple-m1

kitaisreal commented:
@changvvb, @traceon thanks for the job. Great contribution.

nvartolomei commented:
@changvvb can you share some numbers for compilation durations? Wondering if it is practical to use an M1 laptop as a development machine for ClickHouse.

traceon commented:
@nvartolomei

MacBook Air M1, 8 CPU + 8 GPU, 16 GB RAM

---

 AppleClang + Ninja toolchain:

`cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo ..`
14,52s user 8,40s system 77% cpu 29,555 total

`cmake --build . --config RelWithDebInfo --parallel`
9417,54s user 503,73s system 537% cpu 30:46,89 total

---

 AppleClang + Xcode toolchain:

XCODE_IDE=1 cmake -G Xcode -DCMAKE_BUILD_TYPE=RelWithDebInfo ..
106,40s user 45,91s system 39% cpu 6:26,60 total

Building ALL_BUILD with RelWithDebInfo from Xcode
27:02 total

---

changvvb commented:
> @changvvb can you share some numbers for compilation durations? 

@nvartolomei It spends about 20 - 30 minutes to  compile the entire project, faster than X86_64.

alexey-milovidov commented:
With Ryzen 3950 CPU you can have compilation time less than 10 minutes, and the total cost of the workstation will be slightly less than MacBook laptop.

traceon commented:
@alexey-milovidov I doubt that anyone buying a MacBook is optimizing for cost/cpu performance only.
